lib/lure/repo.go

```diff
@@ -183,3 +190,132 @@ func Execute(pwd string, command string, params ...string) (string, error) {
 
 	return out, nil
 }
+
+func cleanupUpdateBranches(auth Authentication, project Project, repo Repo) error {
+
+	if project.Vcs != Hg {
+		return nil
+	}
+
+	trashBranch := project.TrashBranch
+	if trashBranch == "" {
+		log.Printf("Info: Project has no trash branch defined. Skipping cleanup.")
+		return nil
+	}
+
+	log.Printf("Info: Cleaning up lure branches with no associated PRs.")
+
+	out, err := repo.Cmd("branches", "--active", "--template", "{branches}\n")
+	if err != nil {
+		return err
+	}
+
+	branches := strings.Split(out, "\n")
+	branchPrefix := project.BranchPrefix
+	if branchPrefix == "" {
+		branchPrefix = "lure-"
+	}
+	msgRegex, err := regexp.Compile(`Update (\S*) to (\S*)`)
+	if err != nil {
+		return err
+	}
+
+	existingPRs := getPullRequests(auth, project.Owner, project.Name, true)
+
+	for _, branch := range branches {
+		if strings.HasPrefix(branch, branchPrefix) {
+
+			dead, err := isBranchDead(repo, branch, existingPRs, msgRegex)
+			if err != nil {
+				continue
+			}
+
+			if dead {
+				closeDeadBranch(repo, branch, trashBranch)
+			}
+		}
+	}
+
+	if os.Getenv("DRY_RUN") == "1" {
+		log.Println("Running in DryRun mode, not doing the pull request nor pushing the changes")
+	} else {
+		if _, err := repo.Push(); err != nil {
+			return err
+		}
+	}
+
+	log.Printf("Info: Lure branches clean up done.")
+
+	return nil
+}
+
+func isBranchDead(repo Repo, branch string, existingPRs []PullRequest, msgRegex *regexp.Regexp) (bool, error) {
+
+	// Get dependency name and version from latest commit message
+	message, err := repo.Cmd("log", "--limit", "1", "--branch", branch, "--template", "{desc}")
+	if err != nil {
+		return false, err
+	}
+
+	matches := msgRegex.FindStringSubmatch(message)
+	if len(matches) != 3 {
+		return false, err
+	}
+	dependencyName := matches[1]
+	dependencyVersion := matches[2]
+
+	// Look for a matching opened PR
+	titleSuffix := fmt.Sprintf("dependency %s to version %s", dependencyName, dependencyVersion)
+	foundPR := false
+	for _, pr := range existingPRs {
+		if pr.State == "OPEN" && strings.Contains(pr.Title, titleSuffix) {
+			foundPR = true
+			break
+		}
+	}
+
+	return !foundPR, nil
+}
+
+func closeDeadBranch(repo Repo, branch string, trashBranch string) error {
```

Comment:
while we're in the repo file (which is tightly coupled with bitbucket for now) we might be able to get away with closing branches by simply calling the DELETE REST endpoint on the branches (https://developer.atlassian.com/bitbucket/api/2/reference/resource/repositories/%7Busername%7D/%7Brepo_slug%7D/refs/branches/%7Bname%7D)

Reply:
while this is a possibility for closing the branch, it won't merge back the branch into the trashBranch

Reply:
you're the hg expert here :)